File: 18.4-sum.py
#
# @lc app=leetcode id=18 lang=python3
#
# [18] 4Sum
#

# @lc code=start
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()

        nums_size = len(nums)
        res = []

        for i in range(nums_size - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            for j in range(i + 1, nums_size - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue

                left = j + 1
                right = nums_size - 1

                while left < right:

                    curr_res = nums[i] + nums[j] + nums[left] + nums[right] - target

                    if curr_res == 0:
                        res.append([nums[i], nums[j], nums[left], nums[right]])
                        left += 1
                        right -= 1 
                        while left < right and nums[left] == nums[left - 1]:
                            left += 1
                        while left < right and nums[right] == nums[right + 1]:
                            right -= 1

                    elif curr_res < 0:
                        left += 1
                    else:
                        right -= 1
        
        return res

            

        
        
        
# @lc code=end

File: 18.4-sum.java
/*
 * @lc app=leetcode id=18 lang=java
 *
 * [18] 4Sum
 */

// @lc code=start
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);

        List<List<Integer>> res = new ArrayList();
        int numsSize = nums.length;
        
        for (int i = 0; i < numsSize - 3; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            for (int j = i + 1; j < numsSize - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) {
                    continue;
                }

                int left = j + 1;
                int right = numsSize - 1;

                while (left < right) {

                    long currentRes = (long) nums[i] + (long) nums[j] + (long) nums[left] + (long) nums[right] - (long) target;

                    if (currentRes == 0) {
                        res.add(List.of(nums[i], nums[j], nums[left], nums[right]));

                        left++;
                        right--;
                        while (left < right && nums[left] == nums[left - 1]) {
                            left++;
                        }

                        while (left < right && nums[right] == nums[right + 1]) {
                            right--;
                        }
                    } else if (currentRes > 0) {
                        right--;
                    } else {
                        left++;
                    }
                }
            }
        }

        return res;

    }
}
// @lc code=end

